IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m

```diff
@@ -26,21 +26,375 @@
+        {
+            [_spinnerTimer invalidate];
+        }
+        _spinnerTimer = [NSTimer scheduledTimerWithTimeInterval:2.0
+                                                         target:self
+                                                       selector:@selector(onStartLoadingIndicator:)
+                                                       userInfo:nil
+                                                        repeats:NO];
+        [_spinnerTimer setTolerance:0.3];
+    }
+}
+
+- (void)webView:(WKWebView *)webView didFinishNavigation:(null_unspecified WKNavigation *)navigation
+{
+    NSURL *url = webView.URL;
+    MSID_LOG_VERBOSE(self.context, @"-didFinishNavigation host: %@", [MSIDAuthority isKnownHost:url] ? url.host : @"unknown host");
+    MSID_LOG_VERBOSE_PII(self.context, @"-didFinishNavigation host: %@", url.host);
+    
+    [self stopSpinner];
+}
+
+- (void)webView:(WKWebView *)webView didFailNavigation:(null_unspecified WKNavigation *)navigation withError:(NSError *)error
+{
+    [self webAuthFailWithError:error];
+}
+
+- (void)webView:(WKWebView *)webView didReceiveAuthenticationChallenge:(NSURLAuthenticationChallenge *)challenge completionHandler:(ChallengeCompletionHandler)completionHandler
+{
+    NSString *authMethod = [challenge.protectionSpace.authenticationMethod lowercaseString];
+    
+    MSID_LOG_VERBOSE(self.context,
+                     @"%@ - %@. Previous challenge failure count: %ld",
+                     @"webView:didReceiveAuthenticationChallenge:completionHandler",
+                     authMethod, (long)challenge.previousFailureCount);
+    
+    [MSIDChallengeHandler handleChallenge:challenge
+                                  webview:webView
+                                  context:self.context
+                        completionHandler:completionHandler];
+}
+
+- (void)webAuthDidCompleteWithURL:(NSURL *)endURL
+{
+    MSID_LOG_INFO(self.context, @"-webAuthDidCompleteWithURL: %@", [MSIDAuthority isKnownHost:endURL] ? endURL.host : @"unknown host");
+    MSID_LOG_INFO_PII(self.context, @"-webAuthDidCompleteWithURL: %@", endURL);
+    
+    [self endWebAuthenticationWithError:nil orURL:endURL];
+}
+
+// Authentication failed somewhere
+- (void)webAuthFailWithError:(NSError *)error
 {
+    if (self.complete)
+    {
+        return;
+    }
+    
+    // Ignore WebKitError 102 for OAuth 2.0 flow.
+    if ([error.domain isEqualToString:@"WebKitErrorDomain"] && error.code == 102)
+    {
+        return;
+    }
+    
+    [self stopSpinner];
+    
+    if([error.domain isEqual:@"WebKitErrorDomain"])
+    {
+        return;
+    }
+    
+    MSID_LOG_ERROR(self.context, @"-webAuthFailWithError error code %ld", (long)error.code);
+    MSID_LOG_ERROR_PII(self.context, @"-webAuthFailWithError: %@", error);
     
+    [self endWebAuthenticationWithError:error orURL:nil];
+}
+
+#pragma mark - Loading Indicator
+
+- (void)onStartLoadingIndicator:(id)sender
+{
+    (void)sender;
+    
+    if (self.loading)
```

Comment:
who sets self.loading?

Reply:
in `- (void)webView:(WKWebView *)webView didStartProvisionalNavigation:(null_unspecified WKNavigation *)navigation`